### PR TITLE
Replace `madmin` SDK with `mc` CLI shim

### PR DESCRIPTION
**Pull request by masnax**
Removes the `madmin` dependency in LXD, and replaces it with a wrapper around the `mc` command line tool for interacting with the `minio` server admin API.

Not much was changed about how LXD interacts with minio, other than the `WaitReady` function where we now need to repeatedly try connecting, because the `NewAdminClient` helper will attempt to register an alias, and thus will error out if the server is not running yet.

### CODE REVIEW

**Comment by masnax** [comment]
@tomponline I'm assuming we'll also need `minio` and `mc` included in the github workflows as well as `lxd-ci`?

**Review by tomponline — CHANGES_REQUESTED** [review]
static analysis failing

**Comment by tomponline** [comment]
static analysis failing

> @tomponline I'm assuming we'll also need `minio` and `mc` included in the github workflows as well as `lxd-ci`?

Yes, minio is already included though.

File: lxd/storage/s3/miniod/admin.go
Comment (tomponline) [inline lxd/storage/s3/miniod/admin.go:54]: Could we have a constant for "lxd-minio" as you've used it in 2 places here, do they represent the same thing?

File: lxd/storage/s3/miniod/admin.go
Comment (tomponline) [inline lxd/storage/s3/miniod/admin.go:115]: According to the docs the file it exports to is ALIAS-iam-metadata.zip

https://min.io/docs/minio/linux/reference/minio-mc-admin/mc-admin-cluster-iam-export.html

File: lxd/storage/s3/miniod/miniod.go
Comment (tomponline) [inline lxd/storage/s3/miniod/miniod.go:130]: The timeout should be part of the ctx passed in right, rather than a separate argument?

File: lxd/storage/s3/miniod/admin.go
Comment (tomponline) [inline lxd/storage/s3/miniod/admin.go:54]: What does this do? Can we instead provide target argument to each invocation rather than requiring client state? Where is that stored?

Comment (masnax) [reply lxd/storage/s3/miniod/admin.go:54]: It's the same as setting a LXD remote, in principle. The CLI seems to work exclusively with these aliases as identifiers, but there is a `--config-dir` option that lets us control the location of the state.

Comment (tomponline) [reply lxd/storage/s3/miniod/admin.go:54]: > Where is that stored?

Comment (tomponline) [reply lxd/storage/s3/miniod/admin.go:54]: as mc is being called inside the snap

File: lxd/storage/s3/miniod/admin.go
Comment (masnax) [reply lxd/storage/s3/miniod/admin.go:115]: I still need to append the alias, but the actual path is `%s-iam-info.zip`

https://github.com/minio/mc/blob/669cb0a9a475e4f66d837a305800d3eb8ae784a1/cmd/admin-cluster-iam-export.go#L99

File: lxd/storage/s3/miniod/admin.go
Comment (masnax) [reply lxd/storage/s3/miniod/admin.go:54]: It does seem to be smart enough to know to set up a separate state directory inside the snap (the alias isn't registered when running the command from the host), but I think we should be explicit about it anyway.

Comment (masnax) [reply lxd/storage/s3/miniod/admin.go:54]: It seems to drop it in `/tmp/.mc` within the snap

Comment (tomponline) [reply lxd/storage/s3/miniod/admin.go:54]: thats probably ok for the snap, but running lxd outside of the snap now means lxd would modify the existing root user's mc aliases?

Comment (tomponline) [reply lxd/storage/s3/miniod/admin.go:54]: > It does seem to be smart enough to know to set up a separate state directory inside the snap (the alias isn't registered when running the command from the host), but I think we should be explicit about it anyway.

Yes i think we should be using a path inside lxd's path rather than potentially modifying an external user's aliases (in the case of non-snap installs)

Comment (masnax) [reply lxd/storage/s3/miniod/admin.go]: I've set the state directory to `/var/snap/lxd/common/lxd/minio`

**Comment by masnax** [comment]
Looks like the V2 CLI doesn't support the V1 server so I've changed the workflow to pull both from latest. Since the binaries will be user specified going forward, I think it's fine to test against the latest release.

File: lxd/storage/s3/miniod/admin.go
Comment (tomponline) [reply lxd/storage/s3/miniod/admin.go]: What about non snap installs?

Comment (masnax) [reply lxd/storage/s3/miniod/admin.go]: I'm using `shared.VarPath` (https://github.com/canonical/lxd/blob/c5337275a370457afdc69480ba453ea89cc152cd/shared/util.go#L219-L230)

So the state dir should be whatever the user is using.

Comment (tomponline) [reply lxd/storage/s3/miniod/admin.go]: OK sounds good thanks

**Comment by tomponline** [comment]
@masnax we just need the doc updates now, @ru-fu can you take a look please to suggest where is appropriate to cover setting up externally provided minio tooling. Ta

**Comment by ru-fu** [comment]
> @masnax we just need the doc updates now, @ru-fu can you take a look please to suggest where is appropriate to cover setting up externally provided minio tooling. Ta

Not sure I know enough about where MinIO is used to suggest the correct place, but giving my best guess. :wink: 

If everybody needs to do this, it should go into https://documentation.ubuntu.com/lxd/en/latest/installing/ - both the snap installation and the installation from source (I assume?).
If it's needed only when using storage buckets, it should be a section in here: https://documentation.ubuntu.com/lxd/en/latest/howto/storage_buckets/
Plus a note here: https://documentation.ubuntu.com/lxd/en/latest/explanation/storage/#storage-buckets

File: lxd/storage/s3/miniod/miniod.go
Comment (tomponline) [inline lxd/storage/s3/miniod/miniod.go:146]: This seems more concise and avoids repetition:

```go
adminClient, err := p.AdminClient()
if adminClient != nil {
	_, err = adminClient.GetConfig(ctx)
	if err == nil {
		return nil
	}
}

lastErr = err
time.Sleep(time.Millisecond * 100)
continue
```

**Comment by tomponline** [comment]
> > @masnax we just need the doc updates now, @ru-fu can you take a look please to suggest where is appropriate to cover setting up externally provided minio tooling. Ta
> 
> Not sure I know enough about where MinIO is used to suggest the correct place, but giving my best guess. 😉
> 
> If everybody needs to do this, it should go into https://documentation.ubuntu.com/lxd/en/latest/installing/ - both the snap installation and the installation from source (I assume?). If it's needed only when using storage buckets, it should be a section in here: https://documentation.ubuntu.com/lxd/en/latest/howto/storage_buckets/ Plus a note here: https://documentation.ubuntu.com/lxd/en/latest/explanation/storage/#storage-buckets

Its only for local storage buckets (not cephobject).

**Comment by ru-fu** [comment]
> Its only for local storage buckets (not cephobject).

OK, then it should be a section in the how-to ("Install requirements for local storage buckets" or "Set up your environment for local storage buckets" or something like this), and a short note in the explanation.

**Comment by tomponline** [comment]
@masnax ready for review?

File: lxd/storage/s3/miniod/miniod.go
Comment (tomponline) [reply lxd/storage/s3/miniod/miniod.go:146]: This is still open.

**Comment by masnax** [comment]
@tomponline Should be good to go now, just added a line to the docs mentioning that the specified path shouldn't be under any user's home directory.

**Comment by masnax** [comment]
Guess I ran the tests too many times, doc tests are failing with a 429 (too many requests)

**Review by tomponline — DISMISSED** [review]

**Comment by tomponline** [comment]
I'll wait for @ru-fu approval

File: doc/howto/storage_buckets.md
Comment (ru-fu) [inline doc/howto/storage_buckets.md:11]:
```suggestion
LXD uses [MinIO](https://min.io) to set up local storage buckets. To use this feature with LXD, you must install both the server and client binaries.
```

File: doc/howto/storage_buckets.md
Comment (ru-fu) [inline doc/howto/storage_buckets.md:16]:
```suggestion
   - Direct download for various architectures:
```

File: doc/howto/storage_buckets.md
Comment (ru-fu) [inline doc/howto/storage_buckets.md:25]:
```suggestion
   - Direct download for various architectures:
```

File: doc/howto/storage_buckets.md
Comment (ru-fu) [inline doc/howto/storage_buckets.md:31]:
```suggestion
If LXD is installed from a snap, you must configure the snap environment to detect the binaries, and restart LXD.
```

File: doc/howto/storage_buckets.md
Comment (ru-fu) [inline doc/howto/storage_buckets.md:37]:
```suggestion
If LXD is installed from another source, both binaries must be included in the `$PATH` that LXD was started with.
```

**Review by ru-fu — DISMISSED** [review]
Some nitpicks for consistency and simpler language, but those can be done later as well.

**Comment by ru-fu** [comment]
crio.org seems to be down right now, so shouldn't be an issue to merge despite the link check failing - it's not because of this PR.

**Comment by tomponline** [comment]
needs a rebase and a few doc tweaks and then is good to be merged.

**Review by tomponline — APPROVED** [review]

**Pull request merged**
